cluster-sharding/src/main/scala/akka/kafka/cluster/sharding/KafkaClusterSharding.scala

```diff
@@ -0,0 +1,231 @@
+   * accordingly so that entities are (eventually) routed to the local Akka cluster member.
+   *
+   * Returns an Akka classic [[akka.actor.ActorRef]] that can be passed to an Alpakka Kafka [[ConsumerSettings]].
+   */
+  @ApiMayChange(issue = "https://github.com/akka/alpakka-kafka/issues/1074")
+  def rebalanceListener(otherSystem: ActorSystem, typeKeyName: String): ActorRef = {
+    val num = shardingRebalanceListenerActorNum.getAndIncrement()
+    otherSystem
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(Props(new RebalanceListener(typeKeyName)), s"kafka-cluster-sharding-rebalance-listener-$num")
+  }
+}
+
+object KafkaClusterSharding extends ExtensionId[KafkaClusterSharding] {
+  @InternalApi
+  sealed trait KafkaClusterShardingContract {
+    def kafkaPartitions: Int
+    def shardId(entityId: String): String = {
+      // simplified version of Kafka's `DefaultPartitioner` implementation
+      val partition = org.apache.kafka.common.utils.Utils
+          .toPositive(Utils.murmur2(entityId.getBytes())) % kafkaPartitions
+      partition.toString
+    }
+  }
+
+  @InternalApi
+  final class KafkaShardingMessageExtractor[M] private[kafka] (val kafkaPartitions: Int)
+      extends ShardingMessageExtractor[ShardingEnvelope[M], M]
+      with KafkaClusterShardingContract {
+    override def entityId(envelope: ShardingEnvelope[M]): String = envelope.entityId
+    override def unwrapMessage(envelope: ShardingEnvelope[M]): M = envelope.message
+  }
+
+  @InternalApi
+  final class KafkaShardingNoEnvelopeExtractor[M] private[kafka] (val kafkaPartitions: Int,
+                                                                  entityIdExtractor: M => String)
+      extends ShardingMessageExtractor[M, M]
+      with KafkaClusterShardingContract {
+    override def entityId(message: M): String = entityIdExtractor(message)
+    override def unwrapMessage(message: M): M = message
+  }
+
+  @InternalApi
+  private[kafka] final class RebalanceListener(typeKeyName: String) extends Actor {
+    private val log = Logging(context.system, this)
+    private val shardAllocationClient = ExternalShardAllocation(context.system).clientFor(typeKeyName)
+    private val address: Address = Cluster(context.system.toTyped).selfMember.address
+
+    override def receive: Receive = {
+      case TopicPartitionsAssigned(_, partitions) =>
+        implicit val ec: ExecutionContextExecutor = context.classicActorContext.dispatcher
+        val partitionsList = partitions.mkString(",")
+        log.info("Consumer group '{}' is assigning topic partitions to cluster member '{}': [{}]",
+                 typeKeyName,
+                 address,
+                 partitionsList)
+        val updates = partitions.map { tp =>
+          val shardId = tp.partition().toString
+          // the Kafka partition number becomes the akka shard id
+          // TODO: Should the shard allocation client support assigning more than 1 shard id at once?
```

Comment:
Yes i think so: https://github.com/akka/akka/issues/28696

Reply:
I updated to TODO to point to this issue.